pwnlib/libcdb.py

```diff
@@ -524,29 +609,74 @@ def search_by_symbol_offsets(symbols, select_index=None, unstrip=True, return_as
 
     params = {'symbols': symbols}
     log.debug('Request: %s', params)
-    matching_libcs = query_libc_rip(params)
-    log.debug('Result: %s', matching_libcs)
+
+    offline_matching = find_local_libc(params)
+    online_matching = query_libc_rip(params) if not offline else []
+
+    if offline_matching is None:
+        offline_matching = []
+    if online_matching is None:
+        online_matching = []
+
+    offline_id = [x["id"] for x in offline_matching] if offline_matching else []
+    online_id = [x["id"] for x in online_matching] if online_matching else []
+
+    # Combine offline and online search results.
+    matching_id = []
+    matching_libcs = []
+    for x in offline_matching + online_matching:
+        if x["id"] not in matching_id:
+            matching_id.append(x["id"])
+            matching_libcs.append(x)
+
+    log.debug('Offline search result: %s', offline_matching)
+    if not offline:
+        log.debug('Online search result: %s', online_matching)
+
+    if raw:
+        return matching_libcs
+
     if matching_libcs is None or len(matching_libcs) == 0:
-        log.warn_once("No matching libc for symbols %r on libc.rip", symbols)
+        log.warn_once("No matching libc for symbols %r", symbols)
         return None
 
     if return_as_list:
         return [libc['buildid'] for libc in matching_libcs]
 
+    matched_libc = None
+
     if len(matching_libcs) == 1:
-        return search_by_build_id(matching_libcs[0]['buildid'], unstrip=unstrip)
+        matched_libc = matching_libcs[0]
 
-    if select_index is not None:
+    if select_index is not None and matched_libc is None:
         if select_index > 0 and select_index <= len(matching_libcs):
-            return search_by_build_id(matching_libcs[select_index - 1]['buildid'], unstrip=unstrip)
+            matched_libc = matching_libcs[select_index - 1]
         else:
             log.error('Invalid selected libc index. %d is not in the range of 1-%d.', select_index, len(matching_libcs))
             return None
 
-    selected_libc = _handle_multiple_matching_libcs(matching_libcs)
-    return search_by_build_id(selected_libc['buildid'], unstrip=unstrip)
+    if matched_libc is None:
+        matched_libc = _handle_multiple_matching_libcs(matching_libcs)
+
+    # Prioritize using the offline libcdb.
+    if matched_libc["id"] in offline_id:
+        cache, cache_valid = _check_elf_cache('libcdb', matched_libc['buildid'], 'buildid')
```

Comment:
```suggestion
        cache, cache_valid = _check_elf_cache('libcdb', matched_libc['buildid'], 'build_id')
```

Another question: what is the difference between using "build_id" and "buildid" in `hash_type`?

Before I submitted this PR, these two strings were mixed, which easily confused me. :(

For example, while passing between functions, it is "build_id", but in online providers and command-line parameters, it is "buildid". This has led to some redundant checks on my part.



- https://github.com/Gallopsled/pwntools/blob/dev/pwnlib/libcdb.py#L81-L82
- https://github.com/Gallopsled/pwntools/blob/dev/pwnlib/commandline/libcdb.py#L77

Reply:
It is how things have grown over time.. [libcdb](https://gitlab.com/libcdb/libcdb) urls use `build_id` while libc.rip uses `buildid` and libc.rip support came later.